crates/items/src/item_stack.rs

```diff
@@ -0,0 +1,262 @@
+#![forbid(unsafe_code, warnings)]
+
+use crate::{Enchantment, Item};
+use serde::{Deserialize, Serialize};
+
+/// Represents an item stack.
+///
+/// An item stack includes an item type, an amount and a bunch of properties (enchantments, etc.)
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct ItemStack {
+    /// The item type of this `ItemStack`.
+    #[serde(rename = "id")]
+    pub item: Item,
+
+    /// The number of items in the `ItemStack`.
+    #[serde(rename = "Count")]
+    pub count: u32,
+
+    /// The `ItemStack` metadata, containing data such as damage,
+    /// repair cost, enchantments...
+    #[serde(rename = "tag")]
+    pub meta: Option<ItemStackMeta>,
+}
+
+/// Represents the metadata of an `ItemStack`. Contains:
+/// * Item title
+/// * Item lore (Optional)
+/// * Item damage (Optional)
+/// * Item repair cost (Optional)
+/// * Item enchantments
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+pub struct ItemStackMeta {
+    /// The displayed title (name) of the associated `ItemStack`.
+    pub title: String,
+
+    /// The displayed lore of the associated `ItemStack`.
+    pub lore: String,
+
+    /// The damage taken by the `ItemStack`.
+    pub damage: Option<u32>,
+
+    /// The cost of repairing the `ItemStack`.
+    pub repair_cost: Option<u32>,
+
+    /// The enchantments applied to this `ItemStack`.
+    pub enchantments: Vec<Enchantment>,
+}
+
+impl ItemStack {
+    /// Creates a new `ItemStack` with the default name (title)
+    /// no lore, no damage, no repair cost and no enchantments.
+    pub fn new(item: Item, count: u32) -> Self {
+        Self {
+            item,
+            count,
+            meta: Some(ItemStackMeta {
+                title: String::from(item.name()),
+                lore: "".to_string(),
+                damage: None,
+                repair_cost: None,
+                enchantments: vec![],
+            }),
+        }
+    }
+
+    /// Returns whether the given item stack has
+    /// the same type as (but not necessarily the same
+    /// amount as) `self`.
+    pub fn has_same_type(&self, other: &Self) -> bool {
+        self.item == other.item
+    }
+
+    /// Returns whether the given item stack has the same damage
+    /// as `self`.
+    pub fn has_same_damage(&self, other: &Self) -> bool {
+        if let (Some(self_meta), Some(other_meta)) = (self.meta.clone(), other.meta.clone()) {
+            self_meta.damage == other_meta.damage
+        } else {
+            self.meta.is_none() && other.meta.is_none()
+        }
+    }
+
+    /// Returns whether the given `ItemStack` has
+    /// the same count as (but not necessarily the same
+    /// type as) `self`.
+    pub fn has_same_count(&self, other: &Self) -> bool {
+        self.count == other.count
+    }
+
+    /// Returns whether the given `ItemStack` has the same
+    /// type and count as `self`.
```

Comment:
Maybe clarify in the docs that we don't account for enchantments, lore, name, etc.